### PR TITLE
Refactor CMake code related to compiler warnings, enable some additional warnings

### PR DESCRIPTION
**Pull request by SiarheiFedartsou**
# Issue
Doesn't fix anything really important, but allows to better control warnings in the future.

## Tasklist

 - [ ] CHANGELOG.md entry ([How to write a changelog entry](http://keepachangelog.com/en/1.0.0/#how))
 - [ ] update relevant [Wiki pages](https://github.com/Project-OSRM/osrm-backend/wiki)
 - [ ] add tests (see [testing documentation](https://github.com/Project-OSRM/osrm-backend/blob/master/docs/testing.md))
 - [ ] review
 - [ ] adjust for comments
 - [ ] cherry pick to release branch

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: cmake/warnings.cmake [hunk 1]
```diff
@@ -0,0 +1,92 @@
+include (CheckCXXCompilerFlag)
+include (CheckCCompilerFlag)
+
+# Try to add -Wflag if compiler supports it
+macro (add_warning flag)
```
Comment (SiarheiFedartsou) [inline cmake/warnings.cmake:5]: Mainly based on https://github.com/ClickHouse/ClickHouse/blob/5bb40e5856ead49d8ccbfc8177255da00f96f88d/cmake/warnings.cmake#L1

File: include/util/static_graph.hpp [hunk 2]
```diff
@@ -313,8 +313,8 @@ class StaticGraph
     }
 
   protected:
-    NodeIterator number_of_nodes;
-    EdgeIterator number_of_edges;
+    NodeIterator number_of_nodes = 0;
```
Comment (SiarheiFedartsou) [inline include/util/static_graph.hpp:316]: We had `maybe-uninitialized` disabled, not sure if there was really problem here, but looks suspicious.

File: src/util/glibc_workaround.cpp [hunk 3]
```diff
@@ -7,13 +7,17 @@
 // This is needed because libstdc++ itself uses this API - its not
 // just an issue of your code using it, ughhh
 
+#pragma GCC diagnostic push
```
Comment (mjjbell) [inline src/util/glibc_workaround.cpp:10]: We can probably remove this whole file. 
We're not supporting GCC versions this old anymore.

File: cmake/warnings.cmake [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+add_warning(delete-incomplete)
+add_warning(duplicated-cond)
+add_warning(disabled-optimization)
+add_warning(init-self)
+add_warning(bool-compare)
+add_warning(logical-not-parentheses)
+add_warning(logical-op)
+add_warning(maybe-uninitialized)
+add_warning(misleading-indentation)
+add_warning(no-return-local-addr)
```
Comment (mjjbell) [inline cmake/warnings.cmake:71]: Should this be a `no_warning` ?

**Review by mjjbell — APPROVED** [review]

File: cmake/warnings.cmake (hunk 4 above)
Comment (SiarheiFedartsou) [reply cmake/warnings.cmake:71]: It is confusing :) `no-` prefix is part of warning name, i.e. it doesn't mean we are disabling it, I'll add a comment

**Pull request merged**